CMakeLists.txt

```diff
@@ -0,0 +1,154 @@
+# Minimum cmake version required by scikit-build-core: 3.15
+cmake_minimum_required(VERSION 3.15)
+
+set(CLP_FFI_PY_INSTALL_LIBS ON CACHE BOOL "Whether to install the built libraries." FORCE)
+
+if(DEFINED SKBUILD_PROJECT_NAME)
+    set(CLP_FFI_PY
+        ${SKBUILD_PROJECT_NAME}
+        CACHE STRING
+        "Use scikit-build-core's project name."
+        FORCE
+    )
+    set(CLP_FFI_PY_VERSION
+        ${SKBUILD_PROJECT_VERSION}
+        CACHE STRING
+        "Use scikit-build-core's project version."
+        FORCE
+    )
+else()
+    # The project is configured outside of scikit-build-core. Normally this is for build testing
+    # purpose, or loading the project into CLion. We don't install the built library in this config.
+    message(WARNING "Configured outside of scikit-build-core. Library installation is disabled.")
+    set(CLP_FFI_PY_INSTALL_LIBS
+        OFF
+        CACHE BOOL
+        "Disable library installation when building outside of scikit-build-core."
+        FORCE
+    )
+    set(CLP_FFI_PY "clp-ffi-py-local-build" CACHE STRING "Use local build name." FORCE)
+    set(CLP_FFI_PY_VERSION "0.0.0" CACHE STRING "Version info is unused for local build." FORCE)
+    set(CMAKE_EXPORT_COMPILE_COMMANDS
+        ON
+        CACHE BOOL
+        "Enable/Disable output of compile commands during generation."
+        FORCE
+    )
+endif()
+
+project(
+    ${CLP_FFI_PY}
+    LANGUAGES
+        C
+        CXX
+    VERSION ${CLP_FFI_PY_VERSION}
+)
+
+# C-extension modules will be compiled as shared libraries. To enable dynamic linking, all libraries
+# (including the static ones) must be compiled with position-independent-code option.
+set(CMAKE_POSITION_INDEPENDENT_CODE ON CACHE BOOL "Compile libraries using PIC.")
+
+# Find Python interpreter, compiler and development environment.
+find_package(
+    Python
+    REQUIRED
+    COMPONENTS
+        Interpreter
+        Development.Module
+)
+
+set(CLP_FFI_PY_LIB_IR "native" CACHE STRING "The name of IR native library")
+python_add_library(
+    ${CLP_FFI_PY_LIB_IR}
+    MODULE
+    WITH_SOABI
+)
+
+target_compile_features(${CLP_FFI_PY_LIB_IR} PRIVATE cxx_std_20)
+
+set(SRC_ROOT "${CMAKE_CURRENT_SOURCE_DIR}/src" CACHE STRING "Root path of C++ source code")
+set(CLP_FFI_PY_SRC_ROOT
+    "${SRC_ROOT}/clp_ffi_py"
+    CACHE STRING
+    "Root path of clp-ffi-py's C++ source code"
+)
+set(CLP_CORE_ROOT "${SRC_ROOT}/clp/components/core" CACHE STRING "Root path of CLP core")
+
+add_subdirectory(${CLP_CORE_ROOT}/src/clp/string_utils)
+
+# Source files of clp-ffi-core
+# Note: we don't add headers hear since they are not considered to be a part of the project.

```

Comment:
Why aren't they considered part of the project? I know CMake ignores header files, so I would imagine the only reason to add header files is if your IDE needs them to understand the project.

Reply:
The reason to include the source files from clp core for compilation is we don't have clp-ffi-core library yet. Strictly speaking clp-ffi-core is an external library that we use; we're not supposed to modify any source/header files. As a result, the headers shouldn't be a part of the project. This doesn't affect CLion's code navigation, i.e., you can still jump to the declaration of a clp-ffi-core function in the IDE because CLion resolves the include path. But these headers are just like any C++ standard library headers, you're not supposed to make any changes, and CLion will try to tell you they are not a part of the project.